Dyn 3026 code coverage analysis folder

RobertGlobant20 opened the pull request:
### Purpose

Increase the code coverage adding test cases for the Libraries/Analysis folder.
I created several unit tests methods using nUnit for the next classes:

- Label
- SurfaceData
- AnalysisExtensions
- Utils

Also I removed two methods from the SurfaceData.cs file since both have zero references and both are private.
The job DynamoSelfServe ran successfully in master-5.

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang 


### FYIs

@alfredo-pozo

RobertGlobant20 commented:
This is a screenshot about the code coverage for the Analysis project using DotCover command line:

![analysis_code_coverage](https://user-images.githubusercontent.com/61755417/89938058-9279d480-dbdb-11ea-9087-75cfcf54bbe8.png)

mjkkirschner commented:
@RobertGlobant20 did you ever get these methods to actually draw labels into the background preview?

RobertGlobant20 commented:
> @RobertGlobant20 did you ever get these methods to actually draw labels into the background preview?

@mjkkirschner No, I tried to find references to the Label class (or classes that inherit of Label class) and is only used in the Label.cs and the test class that i created, then I think there is no way to used in other places, probably the labels in the background preview is using a different class.

aparajit-pratap commented on test/Libraries/AnalysisTests/LabelTests.cs:
```diff
@@ -0,0 +1,38 @@
+            var package = factory.CreateRenderPackage();
+
+            //Act
+            labelTest.Tessellate(package, new TessellationParameters());
```
What are you testing from the `Tessellate` call? I think the assertions below will pass even if you don't call `Tessellate`, wouldn't they?

RobertGlobant20 replied:
@aparajit-pratap 
The Tessellate method is a [IsVisibleInDynamoLibrary(false)], has no code coverage and is not used (inside the Label class o in other classes, event the whole Label class is not used), I was proposing to delete the whole Label.cs file but Michael said the next "if it's public, we cannot get rid of it as it will break semantic versioning."
You are correct, the assertions are checking the results of Label.ByPointAndString method then they don't depend of calling the  Tessellate method, then is it ok if I just delete the call to the Tessellate method (of course if won't have coverage)?

aparajit-pratap replied:
It looks like the package properties are modified in Tessellate. You can check for those properties after the call to Tessellate.

RobertGlobant20 replied:
@aparajit-pratap this test method was re-written due to some comments that Michael did, please review again.

aparajit-pratap commented on test/Libraries/AnalysisTests/LabelTests.cs:
```diff
@@ -0,0 +1,38 @@
+            labelTest.Tessellate(package, new TessellationParameters());
+
+            //Assert
+            Assert.IsNotNull(labelTest);
```
I think this assertion should be before the call to `Tessellate`. Also, can you check if certain properties of `Label` are set in addition to checking for it to be non-null?

RobertGlobant20 replied:
@aparajit-pratap 
This depends of the previous comment, if I delete the call to Tessellate method there is no need to do it, otherwise I will move it before the call.

aparajit-pratap commented on test/Libraries/AnalysisTests/SurfaceDataTests.cs:
```diff
@@ -0,0 +1,96 @@
+            var sad = SurfaceData.BySurfaceAndPoints(TestSurface(),TestUvs());
+
+            //Assert
+            Assert.NotNull(sad);
```
Unrelated, but what is the difference between `Assert.IsNotNull` and `NotNull`?

RobertGlobant20 replied:
@aparajit-pratap 
I was reading that IsNotNull and NotNull are the same but IsNotNull is compatible to previous versions of nUnit then I updated two lines for using IsNotNull.

commit: https://github.com/DynamoDS/Dynamo/pull/10989/commits/0fbc2034d4f413ca7cfe9ae76c525eacb51be946

aparajit-pratap commented on test/Libraries/AnalysisTests/SurfaceDataTests.cs:
```diff
@@ -0,0 +1,96 @@
+            Assert.Throws<ArgumentNullException>(() => SurfaceData.BySurfaceAndPoints(TestSurface(), null));
+
+            //It will raise an ArgumentNullException when we pass the UV values as empty IEnumerable
+            Assert.Throws<ArgumentException>(() => SurfaceData.BySurfaceAndPoints(TestSurface(), new UV[0]));
```
If it throws `ArgumentNullException`, shouldn't you check for `ArgumentNullException` specifically?

RobertGlobant20 replied:
@aparajit-pratap 
The SurfaceData.BySurfaceAndPoints method throws a ArgumentNullException when the first parameter or second parameter are null but when the IEnumerable<UV> parameter is empty it throws a ArgumentException, the I think is correct in this way.
Let me know if you agree.

mjkkirschner commented:
Hi @RobertGlobant20 - whats odd is that https://github.com/DynamoDS/Dynamo/blob/master/src/Libraries/Analysis/Label.cs#L42
should create a point in the background at least.

RobertGlobant20 commented:
> Hi @RobertGlobant20 - whats odd is that https://github.com/DynamoDS/Dynamo/blob/master/src/Libraries/Analysis/Label.cs#L42
> should create a point in the background at least.

@mjkkirschner The tests for the Label class were rewritten and also I applied changes in the Label.cs and HelixRenderPackage.cs files, please review again.

mjkkirschner commented:
hmm, @RobertGlobant20 it looks like this change causes the mono.sln build to fail - thats because now the analysis tests require sharpDX and WPF references.

I think we have two options.

1. remove the analysis tests from the mono.sln
2. Add those tests that require those references to the wpfvisualizationTests project instead of to the analysisTests project, so the analysis test project does not need these windows only references.

what do you think?

mjkkirschner commented on test/Libraries/AnalysisTests/AnalysisTests.csproj:
```diff
@@ -68,6 +83,10 @@
     </ProjectReference>
+    <ProjectReference Include="..\..\..\src\Libraries\CoreNodeModels\CoreNodeModels.csproj">
+      <Project>{D8262D40-4880-41E4-91E4-AF8F480C8637}</Project>
+      <Name>CoreNodeModels</Name>
```
@RobertGlobant20 - this change is causing atleast some of the 600 failures you are getting.
Change this to `<Private>False<Private>` or in visual studio CopyLocal False, same thing.

This binary is being loaded twice, once from the /nodes folder, and now again from the /bin folder (in your PR). It should only be present in the /nodes folder.

To be safe, I would make all of these dependencies copy Local false, but thats probably not the problem.

RobertGlobant20 replied:
@mjkkirschner I removed the CoreNodeModels reference from the AnalysisTests project, then there shouldn't be any problem now.
Thanks

RobertGlobant20 commented:
> hmm, @RobertGlobant20 it looks like this change causes the mono.sln build to fail - thats because now the analysis tests require sharpDX and WPF references.
> 
> I think we have two options.
> 
> 1. remove the analysis tests from the mono.sln
> 2. Add those tests that require those references to the wpfvisualizationTests project instead of to the analysisTests project, so the analysis test project does not need these windows only references.
> 
> what do you think?

@mjkkirschner I took the option 1, I moved the LabelTests.cs to the DynamoCoreWpfTests project inside a Libraries folder, then I removed the sharpDX and Wpf references from the AnalysisTests project.

RobertGlobant20 commented:
> let's wait for the tests, but otherwise looks good.
@mjkkirschner  The Tests were successful executed.
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/master/
